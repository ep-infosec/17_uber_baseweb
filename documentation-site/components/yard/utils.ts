/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/

import type { TProp } from 'react-view';
import type { TProviderValue } from './provider';

export type TPropValueOverrides = {
  [key: string]: {
    active: boolean;
    style: string;
  };
};

export const countProps = (
  props: { [key: string]: TProp },
  propsConfig: { [key: string]: TProp }
) => {
  let changedProps = 0;
  Object.keys(props).forEach((prop) => {
    if (
      prop !== 'overrides' &&
      props[prop].value !== '' &&
      typeof props[prop].value !== 'undefined' &&
      //@ts-ignore
      props[prop].value !== propsConfig[prop].value
    ) {
      changedProps++;
    }
  });
  return changedProps;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const countOverrides = (overrides: any) => {
  if (!overrides) {
    return 0;
  }
  const existingOverrides = overrides.value ? Object.keys(overrides.value) : [];
  return existingOverrides.filter((key) => overrides.value[key].active).length;
};

export const countThemeValues = (themeState: TProviderValue) => {
  if (!themeState) return 0;
  return Object.keys(themeState).length;
};
