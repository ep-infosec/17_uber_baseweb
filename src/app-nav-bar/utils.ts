/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/

import type { NavItem } from './types';

type GetUniqueIdentifier = (a: NavItem) => string | number;

export function defaultMapItemToNode(item: NavItem) {
  if (__DEV__) {
    if (!item.label) {
      throw Error(
        'There needs to be an unique item.label. You can implement a custom mapping with the mapItemToNode prop.'
      );
    }
  }
  return item.label;
}

function defaultGetUniqueIdentifier(item: NavItem) {
  if (__DEV__) {
    if (!item.label) {
      throw Error(
        'There needs to be an unique item.label. You can implement a custom mapping with the getUniqueIdentifier argument to setItemActive.'
      );
    }
  }
  return item.label;
}

export function mapItemsActive(items: NavItem[], predicate: (a: NavItem) => boolean) {
  return items.map<NavItem>((current) => {
    if (predicate(current)) {
      current.active = true;
    } else {
      current.active = false;
    }

    if (current.children) {
      current.children = mapItemsActive(current.children, predicate);
      if (current.children.some((child) => child.active)) {
        current.active = true;
      }
    }

    return current;
  });
}

export function setItemActive(
  items: NavItem[],
  item: NavItem,
  getUniqueIdentifier: GetUniqueIdentifier = defaultGetUniqueIdentifier
) {
  return mapItemsActive(
    items,
    (current) => getUniqueIdentifier(current) === getUniqueIdentifier(item)
  );
}
