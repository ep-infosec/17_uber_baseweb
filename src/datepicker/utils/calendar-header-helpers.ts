/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import { DEFAULT_MONTHS } from '../constants';

export type Option = {
  id: string;
  label: string;
  disabled?: boolean;
};

type GetMonthItemsArgs = {
  filterMonthsList: number[] | null;
  formatMonthLabel: (a: number) => string;
};

const getDefaultMonthItems = (formatMonthLabel: (a: number) => string) =>
  DEFAULT_MONTHS.map<Option>((month) => ({
    id: month.toString(),
    label: formatMonthLabel(month),
  }));

export const filterMonthItems = (monthItems: Option[], filterList: number[]) =>
  monthItems.map<Option>((month) => {
    if (!filterList.includes(Number(month.id))) {
      return {
        ...month,
        disabled: true,
      };
    }
    return month;
  });

export const getFilteredMonthItems = ({
  filterMonthsList,
  formatMonthLabel,
}: GetMonthItemsArgs) => {
  let monthItems = getDefaultMonthItems(formatMonthLabel);

  if (filterMonthsList) {
    monthItems = filterMonthItems(monthItems, filterMonthsList);
  }

  return monthItems;
};
