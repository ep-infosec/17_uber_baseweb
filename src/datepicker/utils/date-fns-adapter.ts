/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import DateFnsUtils from '@date-io/date-fns';
import type { DateIOAdapter } from './types';

// @ts-expect-error todo(flow->ts) DateIOAdapter types is not exactly correct
const adapter: DateIOAdapter<Date> = new DateFnsUtils({});

export default adapter;
