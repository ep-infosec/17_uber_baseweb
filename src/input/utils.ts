/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import type { InputProps, BaseInputProps, InternalState, SharedProps } from './types';

export function getSharedProps<T>(
  props: BaseInputProps<T> | InputProps,
  state: InternalState
): SharedProps {
  const { disabled, error, positive, adjoined, size, required, resize, readOnly } = props;
  const { isFocused } = state;
  return {
    $isFocused: isFocused,
    $disabled: disabled,
    $error: error,
    $positive: positive,
    // @ts-ignore
    $adjoined: adjoined,
    // @ts-ignore
    $size: size,
    $required: required,
    $resize: resize,
    $isReadOnly: readOnly,
  };
}
