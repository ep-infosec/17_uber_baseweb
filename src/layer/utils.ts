/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import type { TetherPlacement, PopperOffset, NormalizedOffset } from './types';

export function toPopperPlacement(placement: TetherPlacement): string {
  return placement.replace(/(Top|Left)$/, '-start').replace(/(Right|Bottom)$/, '-end');
}

/**
 * Takes the offset passed from popper.js and normalizes it
 */
export function parsePopperOffset(offset: PopperOffset): NormalizedOffset {
  return {
    top: Math.floor(offset.top || 0),
    left: Math.floor(offset.left || 0),
  };
}
