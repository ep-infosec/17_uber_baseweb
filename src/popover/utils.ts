/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
/* eslint-disable import/prefer-default-export */
import { ARROW_SIZE, PLACEMENT } from './constants';
import type { Offset, PopoverPlacement } from './types';

const OPPOSITE_POSITIONS = {
  top: 'bottom',
  bottom: 'top',
  right: 'left',
  left: 'right',
} as const;

/**
 * Returns the opposite of the specified position. Useful for tooltip
 * positioning logic.
 * Examples:
 * top -> bottom
 * left -> right
 */
export function getOppositePosition(position: string): string {
  // @ts-ignore
  return OPPOSITE_POSITIONS[position];
}

/**
 * Determines whether or not the specified position is a vertical one (top or bottom)
 */
export function isVerticalPosition(position: string): boolean {
  return position === 'top' || position === 'bottom';
}

/**
 * Simple utility function for capitalizing the first letter of a string
 */
export function capitalize(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

/**
 * Opposite of function above, converts from Popper.js placement
 * to our placement prop
 */
export function fromPopperPlacement(placement: string): PopoverPlacement | null {
  const popoverPlacement: string = placement
    .replace(/(top|bottom)-start$/, '$1Left')
    .replace(/(top|bottom)-end$/, '$1Right')
    .replace(/(left|right)-start$/, '$1Top')
    .replace(/(left|right)-end$/, '$1Bottom');
  // @ts-ignore
  return PLACEMENT[popoverPlacement] || null;
}

/**
 * Splits something like 'topLeft' to ['top', 'left'] for easier usage
 */
export function splitPlacement(placement: PopoverPlacement): string[] {
  const matches = placement.match(/^([a-z]+)([A-Z][a-z]+)?/) || [];
  return (matches as string[])
    .slice(1, 3)
    .filter(Boolean)
    .map((s) => s.toLowerCase());
}

/**
 * Returns margin styles to add spacing between the popover
 * and its anchor.
 */
export function getPopoverMarginStyles(
  arrowSize: number,
  placement: PopoverPlacement,
  popoverMargin: number
) {
  const [position] = splitPlacement(placement);
  const opposite = getOppositePosition(position);
  if (!opposite) {
    return null;
  }
  const property = `margin${capitalize(opposite)}`;
  return {
    [property]: `${arrowSize + popoverMargin}px`,
  };
}

/**
 * Returns CSS rules for the popover animation start keyframe
 */
export function getStartPosition(
  offset: Offset,
  placement: PopoverPlacement,
  arrowSize: number,
  popoverMargin: number
) {
  offset = { ...offset };
  const [position] = splitPlacement(placement);
  const margin = (arrowSize > 0 ? arrowSize : popoverMargin) * 2;
  if (isVerticalPosition(position)) {
    offset.top += position === 'top' ? margin : -margin;
  } else {
    offset.left += position === 'left' ? margin : -margin;
  }
  return `translate3d(${offset.left}px, ${offset.top}px, 0)`;
}

/**
 * Returns CSS rules for the popover animation end keyframe
 */
export function getEndPosition(offset: Offset) {
  return `translate3d(${offset.left}px, ${offset.top}px, 0)`;
}

/**
 * Returns top/left styles to position the popover arrow
 */
export function getArrowPositionStyles(offsets: Offset, placement: PopoverPlacement) {
  const [position] = splitPlacement(placement);
  const oppositePosition = getOppositePosition(position);
  if (!oppositePosition) {
    return null;
  }

  const alignmentProperty: string = isVerticalPosition(position) ? 'left' : 'top';
  return {
    // @ts-ignore
    [alignmentProperty]: `${offsets[alignmentProperty]}px`,
    [oppositePosition]: `-${ARROW_SIZE - 2}px`,
  };
}
