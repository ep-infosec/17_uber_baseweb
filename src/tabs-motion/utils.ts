/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/

import type * as React from 'react';
import { ORIENTATION, FILL } from './constants';

import type { Fill, Orientation } from './types';

export const getTabId = (uid: string, key: React.Key) => `tabs-${uid}-tab-${key}`;
export const getTabPanelId = (uid: string, key: React.Key) => `tabs-${uid}-tabpanel-${key}`;
export const isHorizontal = (orientation: Orientation) => orientation === ORIENTATION.horizontal;
export const isVertical = (orientation: Orientation) => orientation === ORIENTATION.vertical;
export const isRTL = (direction: string) => direction === 'rtl';
export const isIntrinsic = (fill: Fill) => fill === FILL.intrinsic;
export const isFixed = (fill: Fill) => fill === FILL.fixed;
