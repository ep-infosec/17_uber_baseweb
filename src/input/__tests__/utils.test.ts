/*
Copyright (c) Uber Technologies, Inc.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/
import { getSharedProps } from '../utils';
import { ADJOINED, SIZE } from '..';

test('Utils - getSharedProps', () => {
  const props = {
    id: 'id',
    inputRef: { current: { focus: () => {} } },
    adjoined: ADJOINED.none,
    size: SIZE.default,
    required: false,
    disabled: false,
    readOnly: false,
    error: false,
    positive: false,
    value: 'value',
    placeholder: 'placeholder',
    onChange: jest.fn(),
    autoFocus: false,
    isFocused: false,
  };
  const state = {
    isFocused: true,
  };
  // eslint-disable-next-line jest/no-restricted-matchers
  expect(getSharedProps(props, state)).toMatchSnapshot('getSharedProps returns correct object');
});
