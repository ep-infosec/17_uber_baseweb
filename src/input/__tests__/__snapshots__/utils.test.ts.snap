// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils - getSharedProps: getSharedProps returns correct object 1`] = `
Object {
  "$adjoined": "none",
  "$disabled": false,
  "$error": false,
  "$isFocused": true,
  "$isReadOnly": false,
  "$positive": false,
  "$required": false,
  "$resize": undefined,
  "$size": "default",
}
`;
